### PR TITLE
Networking update - ClusterIP instead of Headless services

### LINKED ISSUE #64: [Networking] Use ClusterIP when possible
# Context
Today, Compose on Kubernetes creates Headless Services to enable Service to Service communication. This 
brings the same developer experience as when working with `docker-compose` or `swarm`, but it is not always the best option. We should prefer creating ClusterIP if we get enough information about how to configure it properly.

## Headless service pros
- No KubeProxy involved: direct POD to POD communication means less latency, less load on the kube-proxy component
- Do not require a list of exposed port: As there is no proxy involved, a headless service alows to reach any port opened by a POD

## Headless service cons
- No real load balancing: the DNS server only answers with as many records as there are PODs matching its selector. It is the responsibility of the client network stack to decide if it should do some round-robin on those DNS entries. Many development runtime don't do that, and worse than that, some just take the first entry into account
- Scalability issue: depending on DNS caching policy on the client, its knowledge of the valid PODs IP can be stale. Worse, some network client stacks have a very aggressive DNS caching policy, and never re-query the DNS server: those can become stuck trying to reach PODs that have been unscheduled due to a change in service scaling.

## Potential solution
Introduce the option to use ClusterIP services instead of headless using this strategy:
- introduce a field on ServiceConfig named `InternalServiceType` that can be `Headless` or `ClusterIP`
- introduce a default behavior, that looks for the `Expose` field in ServiceConfig (see https://docs.docker.com/compose/compose-file/#expose): if it is empty, defaults to Headless. Else use ClusterIP

ClusterIP ports configuration will use the `Expose` field in the ServiceConfig to decide which ports to route.

### PR DESCRIPTION
**Pull request by simonferquel**
fix #64 
This changes the service creation logic to use ClusterIP instead of Headless services for internal communication.
This is done if the InternalPorts collection on a Service is not empty.

### CODE REVIEW

**Comment by simonferquel** [comment]
@chris-crone PTAL (and someone else as well) :)

File: internal/convert/service.go [hunk 1]
```diff
@@ -149,17 +149,41 @@ func toServices(s latest.ServiceConfig, objectMeta metav1.ObjectMeta, labelSelec
 }
 
 // toHeadlessService creates a Kubernetes headless service.
-func toHeadlessService(objectMeta metav1.ObjectMeta, labelSelector map[string]string, original apiv1.Service) *apiv1.Service {
+func toHeadlessService(objectMeta metav1.ObjectMeta, labelSelector map[string]string, original apiv1.Service,
```
Comment (ijc) [inline internal/convert/service.go]: Is the name of this function still accurate?

File: internal/test/builders/stack.go [hunk 2]
```diff
@@ -629,6 +630,16 @@ func WithPort(target uint32, builders ...func(*latest.ServicePortConfig)) func(*
 	}
 }
 
+// WithIternalPort adds an internal port declaration
+func WithIternalPort(port int32, protocol corev1.Protocol) func(*latest.ServiceConfig) {
```
Comment (ijc) [inline internal/test/builders/stack.go]: Typo: "Iternal"→"Internal"

File: internal/convert/service.go [hunk 3]
```diff
@@ -149,17 +149,33 @@ func toServices(s latest.ServiceConfig, objectMeta metav1.ObjectMeta, labelSelec
 }
 
 // toHeadlessService creates a Kubernetes headless service.
-func toHeadlessService(objectMeta metav1.ObjectMeta, labelSelector map[string]string, original apiv1.Service) *apiv1.Service {
+func toHeadlessService(objectMeta metav1.ObjectMeta, labelSelector map[string]string, original apiv1.Service, internalPorts []latest.InternalPort) *apiv1.Service {
```
Comment (chris-crone) [inline internal/convert/service.go]: We should rename this to something like `toInternalService` as it's not necessarily a `HeadlessService` any more.

File: internal/convert/service.go [hunk 4]
```diff
@@ -148,18 +148,42 @@ func toServices(s latest.ServiceConfig, objectMeta metav1.ObjectMeta, labelSelec
 	return hl, published, random
 }
 
-// toHeadlessService creates a Kubernetes headless service.
-func toHeadlessService(objectMeta metav1.ObjectMeta, labelSelector map[string]string, original apiv1.Service) *apiv1.Service {
+// toInternalService creates a Kubernetes service for service-to-service communication.
```
Comment (chris-crone) [inline internal/convert/service.go]: nit: `s/for service-to-service communication./for intra-stack communication/`

**Pull request merged**